lib/topological_inventory/ansible_tower/operations/source.rb

```diff
@@ -10,58 +10,113 @@ class Source
+              break
+            end
+          end
 
-          auth_id = endpoint_authentications.first.id
-          auth = Core::AuthenticationRetriever.new(auth_id, identity).process
-          return STATUS_UNAVAILABLE unless auth
+          is_missing
+        end
+
+        def connection_status
+          return [STATUS_UNAVAILABLE, ERROR_MESSAGES[:endpoint_not_found]] unless endpoint
+          return [STATUS_UNAVAILABLE, ERROR_MESSAGES[:authentication_not_found]] unless authentication
+
+          connection_check
+        end
 
+        def connection_check
           connection = ::TopologicalInventory::AnsibleTower::Connection.new
-          connection = connection.connect(endpoint.host, auth.username, auth.password)
+          connection = connection.connect(endpoint.host, authentication.username, authentication.password)
           connection.api.version
 
-          STATUS_AVAILABLE
+          [STATUS_AVAILABLE, nil]
         rescue => e
           logger.error("Failed to connect to Source id:#{source_id} - #{e.message}")
-          STATUS_UNAVAILABLE
+          [STATUS_UNAVAILABLE, e.message]
+        end
+
+        def update_source_and_endpoint(status, error_message = nil)
+          logger.info("Source#availability_check - updating source [#{source_id}] status [#{status}] message [#{error_message}]")
+
+          update_source(status)
+          update_endpoint(status, error_message)
+        end
+
+        def update_source(status)
+          source = ::SourcesApiClient::Source.new
+          source.availability_status = status
+
+          api_client.update_source(source_id, source)
+        rescue SourcesApiClient::ApiError => e
+          logger.error("Source#availability_check - Failed to update Source id:#{source_id} - #{e.message}")
+        end
+
+        def update_endpoint(status, error_message)
+          if endpoint.nil?
+            logger.error("Source#availability_check - Failed to update Endpoint for Source id:#{source_id}. Endpoint not found")
+            return
+          end
+
+          endpoint_update = SourcesApiClient::Endpoint.new
+
+          endpoint_update.availability_status = status
+          endpoint_update.availability_status_error = error_message # if error_message
+
+          api_client.update_endpoint(endpoint.id, endpoint_update)
+        rescue SourcesApiClient::ApiError => e
+          logger.error("Source#availability_check - Failed to update Endpoint(ID: #{endpoint.id}) - #{e.message}")
+        end
+
+        def endpoint
+          @endpoint ||= api_client.list_source_endpoints(source_id)&.data&.detect(&:default)
+        end
+
+        def authentication
+          return @authentication if @authentication.present?
+
+          endpoint_authentications = api_client.list_endpoint_authentications(endpoint.id.to_s).data || []
+          return nil if endpoint_authentications.empty?
+
+          auth_id = endpoint_authentications.first.id
```

Comment:
I'm just more curious here - it looks like we have a few endpoints that have more than one authentication, found with this query: `Endpoint.all.map(&:authentications).map(&:count).uniq`.

Should we make this loop through the authentications if one fails or is that basically never going to happen? I know it would be difficult, but maybe we could memoize `endpoint_authentications` and if one fails, just change `@authentication` to the next one and `retry`. 

I'm just shooting in the dark here, I'm still unfamiliar with a lot of this but just thought I'd put out my thoughts .

Reply:
@lindgrenj6 that's surprise for me! @syncrou do you know what does it mean? If there are more authentications for one endpoint, which one should be used for operations and collectors?

Reply:
anyway this problem is not related only to tower operations, so I would recommend to merge this PR with this and solved either in following PR in providers-common, or make an separate issue because it's related to all operations and orchestrator

Reply:
Yes, this is most likely an edge case where other sources have multiple auth's for one endpoint, it shouldn't happen with anything we collect.

I'll create an issue on the `sources-api` so we're at least aware of it, it should be fine as far as the collectors goes.

Reply:
It appears there was discussion around handling multiple Authentications per Endpoint. So the modeling is there.  Though there is not currently a way to designate a `current` or `default` authentication.

For this PR `.first` will suffice as it appears the ability is only exposed via the API and not in the UI.